hypertrace-graphql-labels-schema/src/main/java/org/hypertrace/graphql/label/joiner/DefaultLabelJoinerBuilder.java

```diff
@@ -0,0 +1,97 @@
+package org.hypertrace.graphql.label.joiner;
+
+import static java.util.function.Function.identity;
+
+import io.reactivex.rxjava3.core.Observable;
+import io.reactivex.rxjava3.core.Single;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import javax.inject.Inject;
+import lombok.AllArgsConstructor;
+import lombok.experimental.Accessors;
+import lombok.extern.slf4j.Slf4j;
+import org.hypertrace.core.graphql.context.GraphQlRequestContext;
+import org.hypertrace.graphql.label.dao.LabelDao;
+import org.hypertrace.graphql.label.schema.LabelResultSet;
+import org.hypertrace.label.config.service.v1.GetLabelsRequest;
+import org.hypertrace.label.config.service.v1.GetLabelsResponse;
+import org.hypertrace.label.config.service.v1.Label;
+
+@Slf4j
+class DefaultLabelJoinerBuilder implements LabelJoinerBuilder {
+
+  private final LabelDao labelDao;
+
+  @Inject
+  DefaultLabelJoinerBuilder(LabelDao labelDao) {
+    this.labelDao = labelDao;
+  }
+
+  @Override
+  public Single<LabelJoiner> build(GraphQlRequestContext context) {
+    return Single.just(new DefaultLabelJoiner(context));
+  }
+
+  @AllArgsConstructor
+  private class DefaultLabelJoiner implements LabelJoiner {
+    private final GraphQlRequestContext context;
+
+    @Override
+    public <T> Single<Map<String, LabelResultSet>> joinLabels(
+        Collection<T> joinSources, IdGetter<T> idGetter, LabelsGetter<T> labelsGetter) {
+      return labelDao
+          .getLabelsResponse(context, GetLabelsRequest.getDefaultInstance())
```

Comment:
so here's the usage, couldn't this just call `getLabels`  - it both saves us from exposing the server impl here, but also from having to build the server-gql label convert logic a second time - let the dao do it! The responsibility of the joiner is then to take that tenant-wide response and project it into smaller result sets. We could either create our own impl of label result set to do that, or even better expose a build result set method off the existing LabelResponseConverter than accepts labels and puts them into a result set. I prefer the latter (since it keeps the responsibility of label construction in one class), but either way is fine.

Reply:
Fixed